lib/DDGC/Web/Service/ActivityFeed.pm

```diff
@@ -0,0 +1,86 @@
+package DDGC::Web::Service::ActivityFeed;
+
+# ABSTRACT: Community Activity JSON service
+
+=pod
+
+=head1 NAME
+
+DDGC::Web::Service::ActivityFeed - Log / retrieve Community Activity.
+
+=head1 DESCRIPTION
+
+This service logs any activity you care to generate and allows subscription
+to these activities via a key of your choosing.
+
+=head1 MOUNT POINT
+
+This service is expected to be mounted on '/activityfeed.json'
+
+=head1 REQUEST HANDLERS
+
+=cut
+
+use DDGC::Base::Web::Service;
+use Try::Tiny;
+use List::MoreUtils qw/ uniq /;
+
+sub pagesize { 40 }
+
+sub activity_page {
+    my ( $params ) = @_;
+    $params->{page} //= 1;
+    $params->{pagesize} //= pagesize;
+    rset('ActivityFeed')->search_rs({
+        ( $params->{filter} )
+            ? ( type => { -like => $params->{filter} =~ s/\*/%/gr } )
+            : (),
+        for_user => { '=' => [ uniq( undef, $params->{user} ) ] }
+    }, {
+        order_by => { -desc => 'me.id' },
+        rows     => $params->{pagesize},
+        page     => $params->{page},
+    });
```

Comment:
You need **some** handle on the data, since you are responsible for informing the template engine... but yeah, it's pretty obviously a breach of separation-of-concerns to have large chunks of query plan hanging about what should be a thin request router.

So the request router has the parameters which inform the query plan, but I don't think it can get much simpler than something like:

```perl
 $rs->for_user                        
    ->order_by( { -desc => 'me.id' } )
    ->rows( $params->{pagesize} )     
    ->page( $params->{page} )         
```

Is the alternative to pass all those in as parameters to a single resultset method? In order to be immediately readable it would need to be a hashref of params anyway, and just as chunky (if not more so).

I think chaining rs methods like the above in the controller is far more explicit / readable and an acceptable level of other-layer-access.

MVC, so simples!

https://duckduckgo.com/?q=mvc&ia=images